README.md
# schedulability

home
: https://hg.sr.ht/~ged/Schedulability

code
: https://hg.sr.ht/~ged/Schedulability

docs
: https://deveiate.org/code/schedulability

github
: https://github.com/ged/schedulability


## Description

Schedulability is a library for describing scheduled time. You can specify one
or more periods of time using a simple syntax, then combine them to describe
more-complex schedules.


## Usage

Schedules are represented with Schedulability::Schedule objects, which are
empty by default:

    schedule = Schedulability::Schedule.new
    # => #<Schedulability::Schedule:0x007ffcf2b982b8 (empty)>

An empty Schedule has no time restrictions, and will match any time.

To specify matching times, you'll need to construct a Schedule with one or more
periods.


### Periods

A schedule is made up of zero or more positive periods, and zero or more
negative periods. A time is within the schedule if at least one positive period
and no negative periods match it.

Periods are specified as a String that contains a comma-separated list of
period descriptions. The string `"never"` can be specified to explicitly create
a schedule which will not match any time.

A period description is of the form

    [!] scale {range [range ...]} [scale {range [range ...]}]

Negative periods begin with a `!`; you may also use `not` or `except` for
readability.

Scale must be one of nine different scales (or their equivalent codes):

    Scale  | Scale | Valid Range Values
           | Code  |
    -------+-------+------------------------------------------------------
    year   |  yr   | n     where n is a 4-digit integer
    month  |  mo   | 1-12  or  jan, feb, mar, apr, may, jun, jul,
           |       |           aug, sep, oct, nov, dec
    week   |  wk   | 1-6
    yday   |  yd   | 1-366
    mday   |  md   | 1-31
    wday   |  wd   | 1-7   or  sun, mon, tue, wed, thu, fri, sat
    hour   |  hr   | 0-23  or  12am 1am-11am 12noon 12pm 1pm-11pm
    minute |  min  | 0-59
    second |  sec  | 0-59

The same scale type may be specified multiple times. Additional scales are
unioned with the ranges defined by previous scales of the same type in the same
sub-period.

A `range` is specified in the form:

    <range value>
or

    <range value>-<range value>

For two-value ranges, the range is defined as the period between the first and
second `range value`s. Scales which are in seconds granularity are exclusive of
their end value, but the rest are inclusive. For example, `hr {9am-5pm}` means
9:00:00 AM until 4:59:59 PM, but `wd {Wed-Sat}` runs until one second before
midnight on Saturday.

If the first value is larger than the second value (e.g. `min {20-10}`), the
range wraps (except when the scale specification is `year`). For example,
`month {9-2}` is the same as specifying `month {1-2 9-12}` or `month {1-2}
month {9-12}` or even `month {Jan-Feb Sep-Dec}`.

The range specified by the single-value specification is implicitly between the
`range value` and its next sequential whole value. For example, `hr {9}` is the
same as specifying `hr {9-10}`, `mday {15}` is the same as `mday {15-16}`, etc.

Neither extra whitespace nor case are significant in a period description.
Scales must be specified either in long form (`year`, `month`, `week`, etc.) or
in code form (`yr`, `mo`, `wk`, etc.). Scale forms may be mixed in a period
statement.

Values for week days and months can be abbreviated to three characters
(`Wednesday` == `Wed`, `September` == `Sep`).


#### Period Examples

<dl>
<dt><code>wd {Mon-Fri} hr {9am-4pm}</code></dt>
<dd>Monday through Friday, 9am to 5pm</dd>

<dt><code>wd {Mon Wed Fri} hr {9am-4pm}, wd{Tue Thu} hr {9am-2pm}</code></dt>
<dd>Monday through Friday, 9:00:00am to 3:59:59pm on Monday, Wednesday, and
	Friday, and 9:00:00am to 1:59:59pm on Tuesday and Thursday</dd>

<dt><code>wk {1 3 5} wd {Mon-Fri} hr {9am-5pm}</code></dt>
<dd>Mon-Fri 9:00:00am-4:59:59pm, on odd weeks in the month</dd>

<dt><code>month {Jan-Feb Nov-Dec}</code></dt>
<dd>During Winter in the northern hemisphere.</dd>

<dt><code>mo {Nov-Feb}</code></dt>
<dd>The same thing (Winter) as a wrapped range.</dd>

<dt><code>not mo {Mar-Oct}</code></dt>
<dd>The same thing (Winter) as a negative range.</dd>

<dt><code>mo {jan feb nov dec}</code></dt>
<dd>Northern Winter as single months</dd>

<dt><code>mo {Jan Feb}, mo {Nov Dec}</code></dt>
<dd>Also Northern Winter.</dd>

<dt><code>mo {Jan Feb} mo {Nov Dec}</code></dt>
<dd>Northern Winter.</dd>

<dt><code>minute { 0-29 }</code></dt>
<dd>The first half of every hour.</dd>

<dt><code>hour { 12am - 12pm }</code></dt>
<dd>During the morning.</dd>

<dt><code>sec {0-4 10-14 20-24 30-34 40-44 50-54}</code></dt>
<dd>Alternating 5-second periods every hour.</dd>

<dt><code>wd {mon wed fri} hr {8am - 5pm}, except day {1}</code></dt>
<dd>Every Monday, Wednesday, and Friday from 8am until 4:59:59 PM, except on
	the first of the month.</dd>

<dt><code>wd {1 3 5 7} min {0-29}, wd {2 4 6} min {30-59}</code></dt>
<dd>Every first half-hour on alternating week days, and the second half-hour the
  rest of the week.</dd>
</dl>


### Schedule Objects

Schedules are immutable after they're created, but they have mutator methods to
allow you to compose the schedule you want by combining them, or by using
mutator methods that return a changed copy of the original:

    weekend = Schedulability::Schedule.parse( "wd {Sat - Sun}" )
    weekdays = Schedulability::Schedule.parse( "wd {Mon - Fri}" )
    work_hours = Schedulability::Schedule.parse( "hour {9am - 5pm}" )
    off_hours = Schedulability::Schedule.parse( "hour {5pm - 9am}" )

    ### Boolean operators
    on_duty = weekdays | work_hours
    off_duty = weekend + ( weekdays | off_hours )
    # -or-
    off_duty = ~on_duty

    ### Exclusivity predicates
    on_duty.overlaps?( off_duty )
    # => false
    on_duty.exclusive?( off_duty )
    # => true

    ### Time predicates
    Time.now
    # => 2015-12-22 12:05:44 -0800
    on_duty.include?( Time.now )
    # => true
    on_duty.now?
    # => true
    off_duty.now?
    # => false

    ### Case equality (=== operator)
    case Time.now
    when on_duty
        send_sms( "Stuff happened." )
    when off_duty
        send_email( "Stuff happened." )
    end


## Prerequisites

* Ruby 2.2.0 or better


## Installation

    $ gem install schedulability


## Contributing

You can check out the current development source with
[Mercurial](https://bitbucket.org/ged/schedulability), or if you prefer Git, via
[its Github mirror](https://github.com/ged/schedulability).

After checking out the source, run:

    $ rake newb

This task will install any missing dependencies, run the tests/specs,
and generate the API documentation.


## Authors

- Michael Granger <ged@faeriemud.org>
- Mahlon E. Smith <mahlon@martini.nu>


## License

This library borrows much of its schedule description syntax and several
implementation strategies from the Time::Period Perl module by Patrick Ryan,
used under the terms of the Perl Artistic License.

> Patrick Ryan <perl@pryan.org> wrote it.
> Paul Boyd <pboyd@cpan.org> fixed a few bugs.
>
> Copyright (c) 1997 Patrick Ryan. All rights reserved. This Perl
> module uses the conditions given by Perl. This module may only
> be distributed and or modified under the conditions given by Perl.

The rest is:

Copyright (c) 2015-2020, Michael Granger and Mahlon E. Smith
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the author/s, nor the names of the project's
  contributors may be used to endorse or promote products derived from this
  software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.




History.md
# Release History for schedulability

---

## v0.5.0 [2020-03-04] Michael Granger <ged@faeriemud.org>

Improvements:

- Update for Ruby 2.7


## v0.4.1 [2018-09-12] Michael Granger <ged@FaerieMUD.org>

Bugfixes:

- Fix more indentation warnings


## v0.4.0 [2018-09-12] Michael Granger <ged@FaerieMUD.org>

Bugfixes:

- Remove mixins only used in the specs.
- Fix indentation warnings
- Fix README examples


## v0.3.0 [2016-07-29] Michael Granger <ged@FaerieMUD.org>

Bugfix:

- Add support for 0th hour and 12am (thanks to
  Meining Lu <meining@getupperhand.com> on Github for the patch)


## v0.2.0 [2016-07-20] Michael Granger <ged@FaerieMUD.org>

Bugfixes:

- fixes for the 0th second or minute, and the 59th second
  (thanks to Jason Rogers <jacaetevha@gmail.com> for the patch)
- Make Schedule#empty? take negative periods into account

Enhancements

- Add missing Schedule#overlaps? and #exclusive predicate methods.
- Add a Schedule#to_s that stringifies them back into schedule descriptions



## v0.1.0 [2015-12-30] Michael Granger <ged@FaerieMUD.org>

First release.  Happy New Year!



spec/helpers.rb
#!/usr/bin/env ruby
# vim: set nosta noet ts=4 sw=4:
# frozen_string_literal: true

BEGIN {
	require 'pathname'
	basedir = Pathname.new( __FILE__ ).dirname.parent

	libdir = basedir + "lib"

	$LOAD_PATH.unshift( libdir.to_s ) unless $LOAD_PATH.include?( libdir.to_s )
}

# SimpleCov test coverage reporting; enable this using the :coverage rake task
if ENV['COVERAGE']
	$stderr.puts "\n\n>>> Enabling coverage report.\n\n"
	require 'simplecov'
	SimpleCov.start do
		add_filter 'spec'
	end
end

require 'schedulability'
require 'loggability/spechelpers'


# Helpers specific to Schedulability specs
module Schedulability::SpecHelpers
end # module Schedulability::SpecHelpers


### Mock with RSpec
RSpec.configure do |config|

	config.mock_with( :rspec ) do |mock|
		mock.syntax = :expect
	end

	config.disable_monkey_patching!
	config.example_status_persistence_file_path = "spec/.status"
	config.filter_run :focus
	config.filter_run_when_matching :focus
	config.order = :random
	config.profile_examples = 5
	config.run_all_when_everything_filtered = true
	config.shared_context_metadata_behavior = :apply_to_host_groups
	config.warnings = true

	config.include( Loggability::SpecHelpers )
	config.include( Schedulability::SpecHelpers )
end



spec/schedulability_spec.rb
# -*- rspec -*-

require_relative 'helpers'

require 'rspec'

require 'schedulability'


RSpec.describe Schedulability do

	# This is just a namespace so there isn't anything to test

end



spec/schedulability/mixins_spec.rb
#!/usr/bin/env rspec -cfd

require_relative '../helpers'

require 'schedulability/mixins'


describe Schedulability, 'mixins' do


	describe Schedulability::TimeRefinements do

		using( described_class )


		describe "used to extend Numeric objects" do

			SECONDS_IN_A_MINUTE    = 60
			SECONDS_IN_AN_HOUR     = SECONDS_IN_A_MINUTE * 60
			SECONDS_IN_A_DAY       = SECONDS_IN_AN_HOUR * 24
			SECONDS_IN_A_WEEK      = SECONDS_IN_A_DAY * 7
			SECONDS_IN_A_FORTNIGHT = SECONDS_IN_A_WEEK * 2
			SECONDS_IN_A_MONTH     = SECONDS_IN_A_DAY * 30
			SECONDS_IN_A_YEAR      = Integer( SECONDS_IN_A_DAY * 365.25 )


			it "can calculate the number of seconds for various units of time" do
				expect( 1.second ).to eq( 1 )
				expect( 14.seconds ).to eq( 14 )

				expect( 1.minute ).to eq( SECONDS_IN_A_MINUTE )
				expect( 18.minutes ).to eq( SECONDS_IN_A_MINUTE * 18 )

				expect( 1.hour ).to eq( SECONDS_IN_AN_HOUR )
				expect( 723.hours ).to eq( SECONDS_IN_AN_HOUR * 723 )

				expect( 1.day ).to eq( SECONDS_IN_A_DAY )
				expect( 3.days ).to eq( SECONDS_IN_A_DAY * 3 )

				expect( 1.week ).to eq( SECONDS_IN_A_WEEK )
				expect( 28.weeks ).to eq( SECONDS_IN_A_WEEK * 28 )

				expect( 1.fortnight ).to eq( SECONDS_IN_A_FORTNIGHT )
				expect( 31.fortnights ).to eq( SECONDS_IN_A_FORTNIGHT * 31 )

				expect( 1.month ).to eq( SECONDS_IN_A_MONTH )
				expect( 67.months ).to eq( SECONDS_IN_A_MONTH * 67 )

				expect( 1.year ).to eq( SECONDS_IN_A_YEAR )
				expect( 13.years ).to eq( SECONDS_IN_A_YEAR * 13 )
			end


			it "can calculate various time offsets" do
				starttime = Time.now

				expect( 1.second.after( starttime ) ).to eq( starttime + 1 )
				expect( 18.seconds.from_now ).to be_within( 10.seconds ).of( starttime + 18 )

				expect( 1.second.before( starttime ) ).to eq( starttime - 1 )
				expect( 3.hours.ago ).to be_within( 10.seconds ).of( starttime - 10800 )
			end

		end


		context "used to extend Time objects" do

			it "makes them aware of whether they're in the future or not" do
				Timecop.freeze do
					time = Time.now
					expect( time.future? ).to be_falsey

					future_time = time + 1
					expect( future_time.future? ).to be_truthy

					past_time = time - 1
					expect( past_time.future? ).to be_falsey
				end
			end


			it "makes them aware of whether they're in the past or not" do
				Timecop.freeze do
					time = Time.now
					expect( time.past? ).to be_falsey

					future_time = time + 1
					expect( future_time.past? ).to be_falsey

					past_time = time - 1
					expect( past_time.past? ).to be_truthy
				end
			end


			it "adds the ability to express themselves as an offset in English" do
				Timecop.freeze do
					expect( 1.second.ago.as_delta ).to eq( 'less than a minute ago' )
					expect( 1.second.from_now.as_delta ).to eq( 'less than a minute from now' )

					expect( 1.minute.ago.as_delta ).to eq( 'a minute ago' )
					expect( 1.minute.from_now.as_delta ).to eq( 'a minute from now' )
					expect( 68.seconds.ago.as_delta ).to eq( 'a minute ago' )
					expect( 68.seconds.from_now.as_delta ).to eq( 'a minute from now' )
					expect( 2.minutes.ago.as_delta ).to eq( '2 minutes ago' )
					expect( 2.minutes.from_now.as_delta ).to eq( '2 minutes from now' )
					expect( 38.minutes.ago.as_delta ).to eq( '38 minutes ago' )
					expect( 38.minutes.from_now.as_delta ).to eq( '38 minutes from now' )

					expect( 1.hour.ago.as_delta ).to eq( 'about an hour ago' )
					expect( 1.hour.from_now.as_delta ).to eq( 'about an hour from now' )
					expect( 75.minutes.ago.as_delta ).to eq( 'about an hour ago' )
					expect( 75.minutes.from_now.as_delta ).to eq( 'about an hour from now' )

					expect( 2.hours.ago.as_delta ).to eq( '2 hours ago' )
					expect( 2.hours.from_now.as_delta ).to eq( '2 hours from now' )
					expect( 14.hours.ago.as_delta ).to eq( '14 hours ago' )
					expect( 14.hours.from_now.as_delta ).to eq( '14 hours from now' )

					expect( 22.hours.ago.as_delta ).to eq( 'about a day ago' )
					expect( 22.hours.from_now.as_delta ).to eq( 'about a day from now' )
					expect( 28.hours.ago.as_delta ).to eq( 'about a day ago' )
					expect( 28.hours.from_now.as_delta ).to eq( 'about a day from now' )

					expect( 36.hours.ago.as_delta ).to eq( '2 days ago' )
					expect( 36.hours.from_now.as_delta ).to eq( '2 days from now' )
					expect( 4.days.ago.as_delta ).to eq( '4 days ago' )
					expect( 4.days.from_now.as_delta ).to eq( '4 days from now' )

					expect( 1.week.ago.as_delta ).to eq( 'about a week ago' )
					expect( 1.week.from_now.as_delta ).to eq( 'about a week from now' )
					expect( 8.days.ago.as_delta ).to eq( 'about a week ago' )
					expect( 8.days.from_now.as_delta ).to eq( 'about a week from now' )

					expect( 15.days.ago.as_delta ).to eq( '2 weeks ago' )
					expect( 15.days.from_now.as_delta ).to eq( '2 weeks from now' )
					expect( 3.weeks.ago.as_delta ).to eq( '3 weeks ago' )
					expect( 3.weeks.from_now.as_delta ).to eq( '3 weeks from now' )

					expect( 1.month.ago.as_delta ).to eq( '4 weeks ago' )
					expect( 1.month.from_now.as_delta ).to eq( '4 weeks from now' )
					expect( 36.days.ago.as_delta ).to eq( '5 weeks ago' )
					expect( 36.days.from_now.as_delta ).to eq( '5 weeks from now' )

					expect( 6.months.ago.as_delta ).to eq( '6 months ago' )
					expect( 6.months.from_now.as_delta ).to eq( '6 months from now' )
					expect( 14.months.ago.as_delta ).to eq( '14 months ago' )
					expect( 14.months.from_now.as_delta ).to eq( '14 months from now' )

					expect( 6.year.ago.as_delta ).to eq( '6 years ago' )
					expect( 6.year.from_now.as_delta ).to eq( '6 years from now' )
					expect( 14.years.ago.as_delta ).to eq( '14 years ago' )
					expect( 14.years.from_now.as_delta ).to eq( '14 years from now' )
				end
			end

		end

	end


end



lib/schedulability/mixins.rb
# -*- ruby -*-
#encoding: utf-8

require 'schedulability' unless defined?( Schedulability )

module Schedulability

	# Functions for time calculations
	module TimeFunctions

		###############
		module_function
		###############

		### Calculate the (approximate) number of seconds that are in +count+ of the
		### given +unit+ of time.
		###
		def calculate_seconds( count, unit )
			return case unit
			when :seconds, :second
				count
			when :minutes, :minute
				count * 60
			when :hours, :hour
				count * 3600
			when :days, :day
				count * 86400
			when :weeks, :week
				count * 604800
			when :fortnights, :fortnight
				count * 1209600
			when :months, :month
				count * 2592000
			when :years, :year
				count * 31557600
			else
				raise ArgumentError, "don't know how to calculate seconds in a %p" % [ unit ]
			end
		end
	end # module TimeFunctions


	# Refinements to Numeric to add time-related convenience methods
	module TimeRefinements
		refine Numeric do

			### Number of seconds (returns receiver unmodified)
			def seconds
				return self
			end
			alias_method :second, :seconds

			### Returns number of seconds in <receiver> minutes
			def minutes
				return TimeFunctions.calculate_seconds( self, :minutes )
			end
			alias_method :minute, :minutes

			### Returns the number of seconds in <receiver> hours
			def hours
				return TimeFunctions.calculate_seconds( self, :hours )
			end
			alias_method :hour, :hours

			### Returns the number of seconds in <receiver> days
			def days
				return TimeFunctions.calculate_seconds( self, :day )
			end
			alias_method :day, :days

			### Return the number of seconds in <receiver> weeks
			def weeks
				return TimeFunctions.calculate_seconds( self, :weeks )
			end
			alias_method :week, :weeks

			### Returns the number of seconds in <receiver> fortnights
			def fortnights
				return TimeFunctions.calculate_seconds( self, :fortnights )
			end
			alias_method :fortnight, :fortnights

			### Returns the number of seconds in <receiver> months (approximate)
			def months
				return TimeFunctions.calculate_seconds( self, :months )
			end
			alias_method :month, :months

			### Returns the number of seconds in <receiver> years (approximate)
			def years
				return TimeFunctions.calculate_seconds( self, :years )
			end
			alias_method :year, :years


			### Returns the Time <receiver> number of seconds before the
			### specified +time+. E.g., 2.hours.before( header.expiration )
			def before( time )
				return time - self
			end


			### Returns the Time <receiver> number of seconds ago. (e.g.,
			### expiration > 2.hours.ago )
			def ago
				return self.before( ::Time.now )
			end


			### Returns the Time <receiver> number of seconds after the given +time+.
			### E.g., 10.minutes.after( header.expiration )
			def after( time )
				return time + self
			end


			### Return a new Time <receiver> number of seconds from now.
			def from_now
				return self.after( ::Time.now )
			end

		end # refine Numeric


		refine Time do

			# Approximate Time Constants (in seconds)
			MINUTES = 60
			HOURS   = 60  * MINUTES
			DAYS    = 24  * HOURS
			WEEKS   = 7   * DAYS
			MONTHS  = 30  * DAYS
			YEARS   = 365.25 * DAYS


			### Returns +true+ if the receiver is a Time in the future.
			def future?
				return self > Time.now
			end


			### Returns +true+ if the receiver is a Time in the past.
			def past?
				return self < Time.now
			end


			### Return a description of the receiving Time object in relation to the current
			### time.
			###
			### Example:
			###
			###    "Saved %s ago." % object.updated_at.as_delta
			def as_delta
				now = Time.now
				if now > self
					seconds = now - self
					return "%s ago" % [ timeperiod(seconds) ]
				else
					seconds = self - now
					return "%s from now" % [ timeperiod(seconds) ]
				end
			end


			### Return a description of +seconds+ as the nearest whole unit of time.
			def timeperiod( seconds )
				return case
					when seconds < MINUTES - 5
						'less than a minute'
					when seconds < 50 * MINUTES
						if seconds <= 89
							"a minute"
						else
							"%d minutes" % [ (seconds.to_f / MINUTES).ceil ]
						end
					when seconds < 90 * MINUTES
						'about an hour'
					when seconds < 18 * HOURS
						"%d hours" % [ (seconds.to_f / HOURS).ceil ]
					when seconds < 30 * HOURS
						'about a day'
					when seconds < WEEKS
						"%d days" % [ (seconds.to_f / DAYS).ceil ]
					when seconds < 2 * WEEKS
						'about a week'
					when seconds < 3 * MONTHS
						"%d weeks" % [ (seconds.to_f / WEEKS).round ]
					when seconds < 18 * MONTHS
						"%d months" % [ (seconds.to_f / MONTHS).ceil ]
					else
						"%d years" % [ (seconds.to_f / YEARS).ceil ]
					end
			end

		end # refine Time

	end # module TimeRefinements

end # module Schedulability



lib/schedulability/parser.rb
# -*- ruby -*-
# frozen_string_literal: true

require 'loggability'
require 'schedulability' unless defined?( Schedulability )


# A collection of parsing functions for Schedulability schedule syntax.
module Schedulability::Parser
	extend Loggability


	log_to :schedulability


	# A Regexp that will match valid period scale codes
	VALID_SCALES = Regexp.union(%w[
		year   yr
		month  mo
		week   wk
		yday   yd
		mday   md
		wday   wd
		hour   hr
		minute min
		second sec
	])

	# Scales that are parsed with exclusive end values.
	EXCLUSIVE_RANGED_SCALES = %i[ hour hr minute min second sec ]

	# The Regexp for matching value periods
	PERIOD_PATTERN = %r:
		(\A|\G\s+) # beginning of the string or the end of the last match
		(?<scale> #{VALID_SCALES} )
		s? # Optional plural sugar
		\s*
		\{
			(?<ranges>.*?)
		\}
	:ix

	# Pattern for matching +hour+-scale values
	TIME_VALUE_PATTERN = /\A(?<hour>\d+)(?<qualifier>am|pm|noon)?\z/i

	# Downcased day-name Arrays
	ABBR_DAYNAMES = Date::ABBR_DAYNAMES.map( &:downcase )
	DAYNAMES = Date::DAYNAMES.map( &:downcase )

	# Downcased month-name Arrays
	ABBR_MONTHNAMES = Date::ABBR_MONTHNAMES.map {|val| val && val.downcase }
	MONTHNAMES = Date::MONTHNAMES.map {|val| val && val.downcase }


	###############
	module_function
	###############

	### Normalize an array of parsed periods into a human readable string.
	def stringify( periods )
		strings = []
		periods.each do |period|
			period_string = []
			period.sort_by{|k, v| k}.each do |scale, ranges|
				range_string = String.new( encoding: 'utf-8' )
				range_string << "%s { " % [ scale.to_s ]

				range_strings = ranges.each_with_object( [] ).each do |range, acc|
					if range.min == range.max
						acc << range.min
					elsif range.exclude_end?
						acc << "%d-%d" % [ range.min, range.max + 1 ]
					else
						acc << "%d-%d" % [ range.min, range.max ]
					end
				end

				range_string << range_strings.join( ' ' ) << " }"
				period_string << range_string
			end
			strings << period_string.join( ' ' )
		end

		return strings.join( ', ' )
	end


	### Scan +expression+ for periods and return them in an Array.
	def extract_periods( expression )
		positive_periods = []
		negative_periods = []

		expression.strip.downcase.split( /\s*,\s*/ ).each do |subexpr|
			hash, negative = self.extract_period( subexpr )
			if negative
				negative_periods << hash
			else
				positive_periods << hash
			end
		end

		return positive_periods, negative_periods
	end


	### Return the specified period +expression+ as a Hash of Ranges keyed by scale.
	def extract_period( expression )
		hash = {}
		scanner = StringScanner.new( expression )

		negative = scanner.skip( /\s*(!|not |except )\s*/ )

		while scanner.scan( PERIOD_PATTERN )
			ranges = scanner[:ranges].strip
			scale = scanner[:scale]

			case scale
			when 'year',   'yr'
				hash[:yr] = self.extract_year_ranges( ranges )
			when 'month',  'mo'
				hash[:mo] = self.extract_month_ranges( ranges )
			when 'week',   'wk'
				hash[:wk] = self.extract_week_ranges( ranges )
			when 'yday',   'yd'
				hash[:yd] = self.extract_yday_ranges( ranges )
			when 'mday',   'md'
				hash[:md] = self.extract_mday_ranges( ranges )
			when 'wday',   'wd'
				hash[:wd] = self.extract_wday_ranges( ranges )
			when 'hour',   'hr'
				hash[:hr] = self.extract_hour_ranges( ranges )
			when 'minute', 'min'
				hash[:min] = self.extract_minute_ranges( ranges )
			when 'second', 'sec'
				hash[:sec] = self.extract_second_ranges( ranges )
			else
				# This should never happen
				raise ArgumentError, "Unhandled scale %p!" % [ scale ]
			end
		end

		unless scanner.eos?
			raise Schedulability::ParseError,
				"malformed schedule (at %d: %p)" % [ scanner.pos, scanner.rest ]
		end

		return hash, negative
	ensure
		scanner.terminate if scanner
	end


	### Return an Array of year integer Ranges for the specified +ranges+ expression.
	def extract_year_ranges( ranges )
		ranges = self.extract_ranges( :year, ranges, 2000, 9999 ) do |val|
			Integer( val )
		end

		if ranges.any? {|rng| rng.end == 9999 }
			raise Schedulability::ParseError, "no support for wrapped year ranges"
		end

		return ranges
	end


	### Return an Array of month Integer Ranges for the specified +ranges+ expression.
	def extract_month_ranges( ranges )
		return self.extract_ranges( :month, ranges, 0, MONTHNAMES.size - 1 ) do |val|
			self.map_integer_value( :month, val, [ABBR_MONTHNAMES, MONTHNAMES] )
		end
	end


	### Return an Array of week-of-month Integer Ranges for the specified +ranges+ expression.
	def extract_week_ranges( ranges )
		return self.extract_ranges( :week, ranges, 1, 5 ) do |val|
			Integer( strip_leading_zeros(val) )
		end
	end


	### Return an Array of day-of-year Integer Ranges for the specified +ranges+ expression.
	def extract_yday_ranges( ranges )
		return self.extract_ranges( :yday, ranges, 1, 366 ) do |val|
			Integer( strip_leading_zeros(val) )
		end
	end


	### Return an Array of day-of-month Integer Ranges for the specified +ranges+ expression.
	def extract_mday_ranges( ranges )
		return self.extract_ranges( :mday, ranges, 0, 31 ) do |val|
			Integer( strip_leading_zeros(val) )
		end
	end


	### Return an Array of weekday Integer Ranges for the specified +ranges+ expression.
	def extract_wday_ranges( ranges )
		return self.extract_ranges( :wday, ranges, 0, DAYNAMES.size - 1 ) do |val|
			self.map_integer_value( :wday, val, [ABBR_DAYNAMES, DAYNAMES] )
		end
	end


	### Return an Array of 24-hour Integer Ranges for the specified +ranges+ expression.
	def extract_hour_ranges( ranges )
		return self.extract_ranges( :hour, ranges, 0, 24 ) do |val|
			self.extract_hour_value( val )
		end
	end


	### Return an Array of Integer minute Ranges for the specified +ranges+ expression.
	def extract_minute_ranges( ranges )
		return self.extract_ranges( :minute, ranges, 0, 60 ) do |val|
			Integer( strip_leading_zeros(val) )
		end
	end


	### Return an Array of Integer second Ranges for the specified +ranges+ expression.
	def extract_second_ranges( ranges )
		return self.extract_ranges( :second, ranges, 0, 60 ) do |val|
			Integer( strip_leading_zeros(val) )
		end
	end


	### Return the integer equivalent of the specified +time_value+.
	def extract_hour_value( time_value )
		unless match = TIME_VALUE_PATTERN.match( time_value )
			raise Schedulability::ParseError, "invalid hour range: %p" % [ time_value ]
		end

		hour, qualifier = match[:hour], match[:qualifier]
		hour = hour.to_i

		if qualifier
			raise Schedulability::RangeError, "invalid hour value: %p" % [ time_value ] if
				hour > 12

			if qualifier == 'am' && hour == 12
				hour = 0
			elsif qualifier == 'pm' && hour < 12
				hour += 12
			end

		else
			raise Schedulability::RangeError, "invalid hour value: %p" % [ time_value ] if
				hour < 0 || hour > 24
		end

		return hour
	end


	### Extract an Array of Ranges from the specified +ranges+ string using the given
	### +index_arrays+ for non-numeric values. Construct the Ranges with the given
	### +minval+/+maxval+ range boundaries.
	def extract_ranges( scale, ranges, minval, maxval )
		exclude_end = EXCLUSIVE_RANGED_SCALES.include?( scale )
		valid_range = Range.new( minval, maxval, exclude_end )

		ints = ranges.split( /(?<!-)\s+(?!-)/ ).flat_map do |range|
			min, max = range.split( /\s*-\s*/, 2 )

			min = yield( min )
			raise Schedulability::ParseError, "invalid %s value: %p" % [ scale, min ] unless
				valid_range.cover?( min )
			next [ min ] unless max

			max = yield( max )
			raise Schedulability::ParseError, "invalid %s value: %p" % [ scale, max ] unless
				valid_range.cover?( max )

			if min > max
				Range.new( minval, max, exclude_end ).to_a +
					Range.new( min, maxval, false ).to_a
			else
				Range.new( min, max, exclude_end ).to_a
			end
		end

		return self.coalesce_ranges( ints, scale )
	end


	### Coalese an Array of non-contiguous Range objects from the specified +ints+ for +scale+.
	def coalesce_ranges( ints, scale )
		exclude_end = EXCLUSIVE_RANGED_SCALES.include?( scale )
		ints.flatten!
		return [] if ints.empty?

		prev = ints[0]
		range_ints = ints.sort.slice_before do |v|
			prev, prev2 = v, prev
			prev2.succ != v
		end

		return range_ints.map do |values|
			last_val = values.last
			last_val += 1 if exclude_end
			Range.new( values.first, last_val, exclude_end )
		end
	end


	### Map a +value+ from a period's range to an Integer, using the specified +index_arrays+
	### if it doesn't look like an integer string.
	def map_integer_value( scale, value, index_arrays )
		return Integer( value ) if value =~ /\A\d+\z/

		unless index = index_arrays.inject( nil ) {|res, ary| res || ary.index(value) }
			expected = "expected one of: %s, %d-%d" % [
				index_arrays.flatten.compact.flatten.join( ', ' ),
				index_arrays.first.index {|val| val },
				index_arrays.first.size - 1
			]
			raise Schedulability::ParseError, "invalid %s value: %p (%s)" %
				[ scale, value, expected ]
		end

		return index
	end


	### Return a copy of the specified +val+ with any leading zeros stripped.
	### If the resulting string is empty, return "0".
	def strip_leading_zeros( val )
		return val.sub( /\A0+(?!$)/, '' )
	end

end # module Schedulability::Parser
